Firestore/core/src/firebase/firestore/core/sync_engine.cc

```diff
@@ -526,32 +529,42 @@ void SyncEngine::UpdateTrackedLimboDocuments(
 
 void SyncEngine::TrackLimboChange(const LimboDocumentChange& limbo_change) {
   const DocumentKey& key = limbo_change.key();
-
-  if (limbo_targets_by_key_.find(key) == limbo_targets_by_key_.end()) {
+  if (active_limbo_targets_by_key_.find(key) ==
+      active_limbo_targets_by_key_.end()) {
     LOG_DEBUG("New document in limbo: %s", key.ToString());
+    enqueued_limbo_resolutions_.push_back(key);
+    PumpEnqueuedLimboResolutions();
+  }
+}
 
+void SyncEngine::PumpEnqueuedLimboResolutions() {
+  while (!enqueued_limbo_resolutions_.empty() &&
+         active_limbo_targets_by_key_.size() <
+             max_concurrent_limbo_resolutions_) {
+    DocumentKey key = enqueued_limbo_resolutions_.front();
+    enqueued_limbo_resolutions_.pop_front();
     TargetId limbo_target_id = target_id_generator_.NextId();
-    Query query(key.path());
-    TargetData target_data(query.ToTarget(), limbo_target_id,
-                           kIrrelevantSequenceNumber,
-                           QueryPurpose::LimboResolution);
-    limbo_resolutions_by_target_.emplace(limbo_target_id, LimboResolution{key});
-    remote_store_->Listen(target_data);
-    limbo_targets_by_key_[key] = limbo_target_id;
+    active_limbo_resolutions_by_target_.emplace(limbo_target_id,
+                                                LimboResolution{key});
+    active_limbo_targets_by_key_[key] = limbo_target_id;
```

Comment:
Is there special reason why `[]` is used here as opposed to `emplace` you used in the above line?

If not, we should be consistent here and use `emplace` in both?

Reply:
So that particular line of code is not new but only changed due to the renaming of the variable `limbo_targets_by_key_` to `active_limbo_targets_by_key_`.  If I understand `emplace()` correctly, there is no benefit to using it over `operator[]` because `limbo_target_id` is just an int and therefore does not benefit from move semantics.  I was also trying to minimize the unrelated changes so if you'd still prefer that I change it to `emplace()` then I will do that in a follow-up PR.

Reply:
I see. I am OK with a separate PR to make it more consistent.

Reply:
For what it's worth, `emplace` on a `map` is constructing the `std::pair` representing the key-value pair, so even though the `limbo_target_id` is trivially copyable, the pair is not because the key is a `DocumentKey`.

This is something of an academic concern though: we've structured our code so that vocabulary types like `DocumentKey` are cheaply copyable. They're immutable values that hide a `std::shared_ptr`, making a copy just a reference count bump. In this case `emplace` vs `operator[]` without moving both make a copy so there's essentially no difference (`emplace` with a moved key does save the copy, but that doesn't work here because the key is used below).

In any case, making these consistent is still worthwhile, and feel free to just inline that into this PR. A separate PR just for that small change is more effort than this change is worth.

Reply:
Done.  It turns out that this was the *only* place where `[]` needed to be changed to `emplace`.